modules/sor/sorV2/lib/pathGraph/pathGraph.ts

```diff
@@ -457,38 +506,74 @@ export class PathGraph {
         return filtered;
     }
 
-    private getLimitAmountSwapForPath(path: PathGraphEdgeData[], swapKind: SwapKind): bigint {
-        let limit = path[path.length - 1].pool.getLimitAmountSwap(
-            path[path.length - 1].tokenIn,
-            path[path.length - 1].tokenOut,
-            swapKind,
-        );
-
-        for (let i = path.length - 2; i >= 0; i--) {
-            const poolLimitExactIn = path[i].pool.getLimitAmountSwap(
-                path[i].tokenIn,
-                path[i].tokenOut,
-                SwapKind.GivenIn,
+    private getLimitAmountSwapForPath(path: PathGraphEdgeData[]): bigint {
+        let limit;
+        if (this.getOperationFromPath(path[path.length - 1]) === PathOperation.AddLiquidity) {
+            limit = MAX_UINT256;
+        } else if (this.getOperationFromPath(path[path.length - 1]) === PathOperation.RemoveLiquidity) {
+            limit = path[path.length - 1].pool.getLimitAmountRemoveLiquidity(
+                path[path.length - 1].tokenIn,
+                path[path.length - 1].tokenOut,
+                RemoveLiquidityKind.SingleTokenExactIn,
             );
-            const poolLimitExactOut = path[i].pool.getLimitAmountSwap(
-                path[i].tokenIn,
-                path[i].tokenOut,
-                SwapKind.GivenOut,
+        } else {
+            limit = path[path.length - 1].pool.getLimitAmountSwap(
+                path[path.length - 1].tokenIn,
+                path[path.length - 1].tokenOut,
+                SwapKind.GivenIn,
             );
+        }
 
-            if (poolLimitExactOut <= limit) {
-                limit = poolLimitExactIn;
-            } else {
-                const pulledLimit = path[i].pool.swapGivenOut(
+        for (let i = path.length - 2; i >= 0; i--) {
+            let limitGivenIn;
+            let limitGivenOut;
+            if (this.getOperationFromPath(path[i]) === PathOperation.Swap) {
+                limitGivenIn = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenIn);
+                limitGivenOut = path[i].pool.getLimitAmountSwap(path[i].tokenIn, path[i].tokenOut, SwapKind.GivenOut);
+                if (limitGivenOut <= limit) {
+                    limit = limitGivenOut;
```

Comment:
This is how it was implemented in the original code.
```suggestion
                    limit = limitGivenIn;
```